Add scriptapi interface to HTTPFetchRequest with security

Jeija opened the pull request:
This is an update / rewrite of #1869.

This pull request enables downloading files or requesting URLs, both POST and GET methods, by providing a simple interface to HTTPFechRequest in Lua.
## Use cases:
- Share creations with others, like buildings, worldedit schemas, mesecons luacontroller code snippets
- Inter-server-communication: Chat, trade and much more
- Update server status on a webpage
- Ingame IRC (with server as wrapper HTTP -> IRC)
- Ingame to real-life interface, enables stuff like https://www.youtube.com/watch?v=jo_GT6a4M90
- Lots of other stuff
## Code

You can use different ways to perform requests:

``` Lua
local http_api = minetest.request_http_api()
local handle = http_api.fetch_async({...})
...
local response = http_api.fetch_async_get(handle)
```

This performs a request asynchronously. A response is provided as soon as the server answer or when the connection times out.

``` Lua
http_api.fetch({...}, function (response)
end)
```

This performs an asynchronous http request. The second argument is a callback function that is called as soon as the server answers or when the connection times out.

``` Lua
local response = http_api.fetch_sync({...})
```

This performs a synchronous request which blocks the server thread, but provides direct results as return value.

Requests must have the following form:

``` Lua
{
    url = "http://example.org",
    timeout = 10, -- optional, in seconds, uses default values if none provided
    -- some more data optional here, see lua_api.txt, e.g. User Agent, POST data, Headers, ...
}
```

Responses will be a table like this one:

``` Lua
{
    succeeded = true,
    timeout = false,
    code = 200,
    data = "response data"
}
```
### Demo mod

You can get a [demo mod](http://mesecons.net/random/curl_secure.zip) that uses HTTP requests. Just read through init.lua and get the nodes from the creative inventory.
## Security

Similar to what mod security does with `minetest.request_insecure_environment()`, mods need to request the HTTP APIs when starting up with `minetest.request_http_api()`. Mods are only granted access to the HTTP APIs if mod security is disabled or if the mods are explicitly whitelisted in `secure.http_mods`. `minetest.request_http_api()` returns a table containing the HTTP functions or nil, if access has been denied.

Please, Minetest Devs, I first suggested these HTTP APIs **three** years ago and they were never merged. Now that the security is hardened and the code is hopefully all cleaned up, I don't see any reason not to merge this. I will be very open for constructive criticism, but please provide clear instructions on what I need to do in order to get this in.
I think many modders are looking forward to these APIs.

Jeija committed:
Fix HTTPFetchRequest performing a GET request if post_data is supplied
Instead, perform a POST request with post_data.

Jeija committed:
Add Lua interface to HTTPFetchRequest

This allows mods to perform both asynchronous and synchronous HTTP
requests. Mods are only granted access to HTTP APIs if either mod
security is disabled or if they are whitelisted in the secure.http_mods
setting.

kilbith commented:
Let's push a new generation of mods forward :+1:

est31 commented:
> Check for updates or install mod updates

That's a bad usecase IMO, we should do this centralized.

Generally I do like this PR's idea, I'll review it later.

Jeija commented:
Yeah true, I just copied that text from the old PR... Will remove this usecase.

C1ffisme commented:
This sounds like a good idea... I can't think of anything I'd use it for now, but maybe later on...

Just as long as Minetest doesn't become an Illegal file transfer program.

I wonder if it's possible to google things from minetest using this code, maybe search a wiki?

:+1:

est31 commented on src/script/lua_api/l_http.cpp:
```diff
@@ -0,0 +1,173 @@
+	req.post_data = getstringfield_default(L, 1, "post_data", "");
+	lua_getfield(L, 1, "useragent");
+	if (lua_isstring(L, -1))
+		req.useragent = getstringfield_default(L, 1, "useragent", "");
```
can it be `user_agent`? Only on the part of the API that's exposed to LUA, no need to refactor the `HTTPFetchRequest` type.

est31 commented on src/script/lua_api/l_http.cpp:
```diff
@@ -0,0 +1,173 @@
+	NO_MAP_LOCK_REQUIRED;
+
+	// If mod security is enabled, mod must be listed in secure.http_mods
+	if (g_settings->getBool("secure.enable_security")) {
```
Can it require to be listed in `secure.http_mods` all the time? Then mod authors won't need to change anything in their documentation when we default to security on.

Jeija commented:
@est31 Just updated the commit with your two suggestions.

est31 commented:
`fetch_sync` shouldn't be added because of speed. HTTP requests usually take tens to hundreds of miliseconds. Disabling all server logic for this amount of time while there is a much faster alternative is bad.

`fetch_async`, I'm not sure about this one either. My main problem with it is that `fetch_async_get` doesn't check the return value of `httpfetch_async_get`. 

Also, `httpfetch_caller_alloc` creates pretty guessable handles, so evil mod code can spy on results from different mod fetch requests by trying different handles. This is btw a common problem with APIs that expose handles, google for "guessable handle", it will give you a long list of bugs.

This can be partially solved by providing an `httpfetch_caller_alloc_secure` method (and using it) that does the same as `httpfetch_caller_alloc`, just performing a (secure) hash function on the value instead of doing `++`, and starting with a (secure-)randomly chosen start value. Also, the handle is typed `long`, but lua's number system can't map all numbers inside a 64 bit range. This is probably no problem for the current `httpfetch_caller_alloc` algorithm, as it usually stays below that boundary, but will be a problem if one uses the approach outlined above, with 64 bit based hashes. So either use 32 bit hashes, or cast the 64 bit hash before returning and doing the check in the map (probably more secure). Also it would be required to migrate to `u32` handles.

So, if `httpfetch_caller_alloc` is worth that effort, and you fix the things I outlined above, I agree to it, or you remove `httpfetch_caller_alloc`, your choice.

The `fetch` callback is fine, I like it.

Jeija committed:
Make HTTP caller ID non-predictable and use http_async_get return value

Add httpfetch_caller_alloc_secure to generate random, non-predictable
caller IDs so that lua mods cannot spy on each others HTTP queries
and use return value of http_async_get to specify if the request
has completed.

rubenwardy commented:
:+1: long overdue

est31 commented on builtin/game/misc.lua:
```diff
@@ -178,3 +178,22 @@ function core.raillike_group(name)
+			if res.completed then
+				callback(res)
+			else
+				core.after(0, update_http_status)
```
That's busy polling, and not good. There should be a better polling strategy done here, e.g. have it check all 10 ms or so.

Jeija committed:
Remove HTTPApiTable.fetch_sync (synchronous HTTP requests)

Makes sure modders can't block the whole server thread.

est31 commented:
Okay for me.

Jeija committed:
Use post_data both for fields and raw POST data

Mods that are listed in secure.trusted_mods may now
also use HTTP requests. Increase default timeout for HTTP
requests to 3 seconds.

nerzhul commented:
:+1: for me too.

est31 commented:
Merged as two commits: a3892f5a6632550bf0c14c18e6902f6ae06bb567 and 31e0667a4a53a238d0321194b57b083bd74c0a5b
